From 7ba26cc6e67b9f7614846b12ab941ede63e03f5c Mon Sep 17 00:00:00 2001
From: Shadi Ammouri <shadi@marvell.com>
Date: Wed, 4 Dec 2013 09:11:59 +0200
Subject: [PATCH 1179/1825] ctrlEnv: Change mvCtrlModelGet() to be more SMP
 friendly

https://github.com/MISL-EBU-System-SW/misl-windriver.git linux-3.4.69-14t2-read
commit e7641f0ec5055efb13f9788e0d30a9f46afb1295

	Updated implementation of mvCtrlModelGet() to read the model
	from the registers only once, and store the value in a global variable
	that will be used for subsequent calls.
	This change is needed to prevent the CPU from being stucked in case this
	function is called by two different CPUs simultaniously, where one of the
	CPUs finishes the function execution (and clock gates the PEX unit) while
	the second CPU is just before reading the device ID.
	This solution is not 100% SMP friendly, and assumes that the first call
	for this function will happen when only a single CPU is running (which is
	a reasonable assumption).
	Issue EBU_System_SW/SYSTEMSW-60

Change-Id: I612adc10637de153e1ab567162bc1e442064dd1e
Signed-off-by: Shadi Ammouri <shadi@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/4582
Reviewed-by: Nadav Haklai <nadavh@marvell.com>
Tested-by: Nadav Haklai <nadavh@marvell.com>
Signed-off-by: Zhong Hongbo <hongbo.zhong@windriver.com>
---
 .../armada_xp_family/ctrlEnv/mvCtrlEnvLib.c        |    5 +++++
 1 files changed, 5 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-armadaxp/armada_xp_family/ctrlEnv/mvCtrlEnvLib.c b/arch/arm/mach-armadaxp/armada_xp_family/ctrlEnv/mvCtrlEnvLib.c
index 3bf6cf5..837b0b8 100755
--- a/arch/arm/mach-armadaxp/armada_xp_family/ctrlEnv/mvCtrlEnvLib.c
+++ b/arch/arm/mach-armadaxp/armada_xp_family/ctrlEnv/mvCtrlEnvLib.c
@@ -829,6 +829,10 @@ MV_U16 mvCtrlModelGet(MV_VOID)
 	MV_U32 devId;
 	MV_U16 model = 0;
 	MV_U32 reg, reg2;
+	static MV_U16 modelId = 0x0;
+
+	if (modelId != 0x0)
+		return modelId;
 
 	/* if PEX0 clocks are disabled - enabled it to read */
 	reg = MV_REG_READ(POWER_MNG_CTRL_REG);
@@ -846,6 +850,7 @@ MV_U16 mvCtrlModelGet(MV_VOID)
 
 	model = (MV_U16) ((devId >> 16) & 0xFFFF);
 
+	modelId = model;
 	return model;
 }
 
-- 
1.7.5.4

